File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    void give(set<vector<int>>& a, vector<int>& c, int t, int s, int i, vector<int>& pa) {
        if (s > t || i == c.size()) {
            return;
        }
        if (s == t) {
            a.insert(pa);
            return; // Ensure no further recursion occurs after finding a valid combination
        }
        pa.push_back(c[i]);
        give(a, c, t, s + c[i], i, pa); // Include the current element
        pa.pop_back();
        give(a, c, t, s, i + 1, pa); // Move to the next element
    }

    vector<vector<int>> combinationSum(vector<int>& c, int t) {
        set<vector<int>> a;
        vector<int> pa;
        give(a, c, t, 0, 0, pa);
        return vector<vector<int>>(a.begin(), a.end()); // Construct vector directly from the set
    }
};

File: 39-combination-sum/combination-sum.go
package main

func give(i, t int, arr []int, a *[][]int, ds []int) {
    if i == len(arr) {
        if t == 0 {
            com := append([]int{}, ds...) // copy ds to avoid reference issues
            *a = append(*a, com)
        }
        return
    }

    if arr[i] <= t {
        ds = append(ds, arr[i])   // choose current element
        give(i, t-arr[i], arr, a, ds) // explore with the current element
        ds = ds[:len(ds)-1]       // backtrack
    }

    give(i+1, t, arr, a, ds)  // explore without the current element
}

func combinationSum(c []int, t int) [][]int {
    var a [][]int
    var ds []int
    give(0, t, c, &a, ds)
    return a
}
